### PR TITLE
Various fixes linked to mc_rtc/topic/TVMBackend

### PR DESCRIPTION
**Pull request by aescande**
This PR introduce various fixes on problem triggered in the mc_rtc use of TVM.
Here are the main ones:
 *  `VariableCountingVector` fully handling non-euclidean variable. In particular, this removes the limitation that the resulting vector of variables should not be differentiated
 * Subvariable substitution was adapted and fixed in consequence
 * Various fixes to use the proper variable size depending on derivative level.

### CODE REVIEW

File: src/solver/LSSOLLeastSquareSolver.cpp [hunk 1]
```diff
@@ -35,6 +35,23 @@ LSSOLLeastSquareSolver::LSSOLLeastSquareSolver(const LSSOLLSSolverOptions & opti
   TVM_PROCESS_OPTION(printLevel, fp_)
   TVM_PROCESS_OPTION(rankTol, fp_)
   TVM_PROCESS_OPTION(warm, fp_)
+  ls_.forceMinSize(false);
```
Comment (aescande) [inline src/solver/LSSOLLeastSquareSolver.cpp:38]: Is `TVM` the proper place to define these default values? Versus defining them in `mc_rtc`?
I don't have strong opinion on this, but it's worth considering the question as we impose these values to all `TVM` uses.

File: src/solver/LSSOLLeastSquareSolver.cpp [hunk 2]
```diff
@@ -35,6 +35,23 @@ LSSOLLeastSquareSolver::LSSOLLeastSquareSolver(const LSSOLLSSolverOptions & opti
...
+  if(!options.feasibilityMaxIter())
+  {
+    ls_.feasibilityMaxIter(4 * ls_.feasibilityMaxIter());
+    fp_.feasibilityMaxIter(4 * fp_.feasibilityMaxIter());
+  }
+  if(!options.optimalityMaxIter())
+  {
+    ls_.optimalityMaxIter(4 * ls_.optimalityMaxIter());
+    fp_.optimalityMaxIter(4 * fp_.optimalityMaxIter());
+  }
```
Comment (aescande) [inline src/solver/LSSOLLeastSquareSolver.cpp:54]: The preferred way to add these values would be to use `TVM_ADD_NON_DEFAULT_OPTION` in the '.h'

Comment (gergondet) [reply src/solver/LSSOLLeastSquareSolver.cpp:54]: Yes. But will it allow to use `4 * ls_.optimialityMaxIter()` instead of an hard-coded value?

**Comment by gergondet** [comment]
@aescande can you open an issue related to [5f44ec1](https://github.com/jrl-umi3218/tvm/pull/37/commits/5f44ec1f9d896bf5d01707ce3874a87c9f59e88a)?

File: src/solver/LSSOLLeastSquareSolver.cpp (hunk 1 above)
Comment (gergondet) [reply src/solver/LSSOLLeastSquareSolver.cpp:38]: The issue here is that mc_rtc uses the default ls solver so it doesn't know what options to pass. That would also prevent drop-in replacement when switching the default solver in TVM

**Comment by gergondet** [comment]
Hi @aescande 

Can we merge this in the current state?

**Pull request merged**